File: src/test/kotlin/school/charset/app/domain/encoding/EncodingTest.kt
package school.charset.app.domain.encoding

import io.kotest.core.spec.style.FreeSpec
import io.kotest.matchers.shouldBe

class EncodingTest :
    FreeSpec({

        "fromId" - {
            "returns the correct encoding for valid IDs" {
                Encoding.fromId("ascii") shouldBe Encoding.Ascii
                Encoding.fromId("latin1") shouldBe Encoding.Latin1
                Encoding.fromId("windows-1252") shouldBe Encoding.Windows1252
                Encoding.fromId("utf-8") shouldBe Encoding.Utf8
                Encoding.fromId("utf-16be") shouldBe Encoding.Utf16Be
                Encoding.fromId("utf-16le") shouldBe Encoding.Utf16Le
                Encoding.fromId("utf-32be") shouldBe Encoding.Utf32Be
                Encoding.fromId("utf-32le") shouldBe Encoding.Utf32Le
            }

            "returns null for invalid IDs" {
                Encoding.fromId("invalid-encoding") shouldBe null
                Encoding.fromId("") shouldBe null
            }
        }
    })

File: test/charset/encoding/encoding_test.exs
defmodule Charset.EncodingTest do
  use ExUnit.Case, async: true

  alias Charset.Encoding

  doctest Charset.Encoding

  describe "from_id/1" do
    test "returns the correct encoding for valid ids" do
      assert Encoding.from_id("ascii") == :ascii
      assert Encoding.from_id("latin1") == :latin1
      assert Encoding.from_id("windows-1252") == :windows1252
      assert Encoding.from_id("utf-8") == :utf8
      assert Encoding.from_id("utf-16be") == :utf16be
      assert Encoding.from_id("utf-16le") == :utf16le
      assert Encoding.from_id("utf-32be") == :utf32be
      assert Encoding.from_id("utf-32le") == :utf32le
    end

    test "returns nil for invalid ids" do
      assert Encoding.from_id("invalid-encoding") == nil
      assert Encoding.from_id("") == nil
    end

    test "roundtrips with id/1 for every encoding" do
      for encoding <- Encoding.all() do
        assert encoding |> Encoding.id() |> Encoding.from_id() == encoding
      end
    end
  end
end
